src/Compilers/Core/Portable/FileSystem/PathUtilities.cs

```diff
@@ -773,6 +775,46 @@ public static bool IsValidFilePath([NotNullWhen(true)] string? fullPath)
         public static string NormalizeWithForwardSlash(string p)
             => DirectorySeparatorChar == '/' ? p : p.Replace(DirectorySeparatorChar, '/');
 
+        /// <summary>
+        /// Takes a normalized full path from <see cref="NormalizeWithForwardSlash(string)"/> and expands any '..' or '.'
+        /// into their equivalent representation.
+        /// </summary>
+        /// <returns>An equivalent path that does not contain any '..' or '.' path parts.</returns>
+        public static string ExpandNormalizedFullPathWithRelativeParts(string p)
+        {
+            if (!IsAbsolute(p))
+            {
+                // if this isn't an absolute path we can't expand it correctly without a root path
+                return p;
+            }
+
+            // GetPathParts also removes any instances of '.'
+            var parts = GetPathParts(p);
+
+            // On windows we need to skip the volume specifier, but remember it for re-joining later
+            var volumeSpecifier = (!IsUnixLikePlatform) ? string.Empty : p.Substring(0, 2);
+
+            // Skip the root directory
+            var toSkip = IsUnixLikePlatform ? 1 : 2;
```

Comment:
>IsUnixLikePlatform ? 1 : 2 [](http://example.com/codeflow?start=25&length=26)

Are we guaranteed there are at least two parts on Windows?

Reply:
It actually would work with UNC paths, but not entirely correctly. Instead i've just made it bail if we see them, as this is supposed to be best effort.